shopping-cart/shopping-cart-scala/shopping-cart/src/main/scala/com/example/shoppingcart/impl/ShoppingCartEntity.scala

```diff
@@ -1,210 +1,149 @@
+    if (items.isEmpty)
+      Effect.reply(cmd)(Rejected("Cannot checkout empty cart"))
+    else
+      Effect
+        .persist(CheckedOut(Instant.now()))
+        .thenReply(cmd) { _ =>
+          Accepted
+        }
 
-      // Command handler for the Hello command
-      case (Get, ctx, state) =>
-        // Reply with the current state.
-        ctx.reply(state)
+  def onReadState(cmd: Get): ReplyEffect[ShoppingCartEvent, ShoppingCartState] =
+    Effect.reply(cmd)(CurrentState(this))
 
-    }.onEvent(eventHandlers)
+  def applyEvent(evt: ShoppingCartEvent): ShoppingCartState = {
+    evt match {
+      case ItemUpdated(productId, quantity, _) => updateItem(productId, quantity)
+      case CheckedOut(_)                       => checkout
+    }
   }
 
-  def checkedOut = {
-    Actions().onReadOnlyCommand[Get.type, ShoppingCartState] {
-
-      // Command handler for the Hello command
-      case (Get, ctx, state) =>
-        // Reply with the current state.
-        ctx.reply(state)
-
-    }.onCommand[UpdateItem, Done] {
+  private def updateItem(productId: String, quantity: Int) = {
+    quantity match {
+      case 0 => copy(items = items - productId)
+      case _ => copy(items = items + (productId -> quantity))
+    }
+  }
 
-      // Not valid when checked out
-      case (_, ctx, _) =>
-        ctx.commandFailed(ShoppingCartException("Can't update item on already checked out shopping cart"))
-        ctx.done
+  private def checkout = copy(checkedOut = true)
+}
 
-    }.onCommand[Checkout.type, Done] {
+object ShoppingCartState {
 
-      // Not valid when checked out
-      case (_, ctx, _) =>
-        ctx.commandFailed(ShoppingCartException("Can't checkout on already checked out shopping cart"))
-        ctx.done
+  def empty: ShoppingCartState = ShoppingCartState(Map.empty, checkedOut = false)
 
-    }.onEvent(eventHandlers)
-  }
+  val typeKey = EntityTypeKey[ShoppingCartCommand[_]]("ShoppingCartEntity")
 
-  def eventHandlers: EventHandler = {
-    // Event handler for the ItemUpdated event
-    case (ItemUpdated(productId: String, quantity: Int, _), state) => state.updateItem(productId, quantity)
+  def behavior(entityContext: EntityContext): Behavior[ShoppingCartCommand[_]] = {
 
-    // Event handler for the checkout event
-    case (_: CheckedOut, state) => state.checkout
-  }
-}
+    val persistenceId = typeKey.persistenceIdFrom(entityContext.entityId)
 
-/**
-  * The current state held by the persistent entity.
-  */
-case class ShoppingCartState(items: Map[String, Int], checkedOut: Boolean) {
+    EventSourcedBehavior
+      .withEnforcedReplies[ShoppingCartCommand[_], ShoppingCartEvent, ShoppingCartState](
+        persistenceId = persistenceId,
+        emptyState = ShoppingCartState.empty,
+        commandHandler = (cart, cmd) => cart.applyCommand(cmd),
+        eventHandler = (cart, evt) => cart.applyEvent(evt)
+      )
+      .withTagger(LagomTaggerAdapter(entityContext, ShoppingCartEvent.Tag))
```

Comment:
Recommended way for existing Lagom users should be to use the existing tagging from Lagom, but adapt the Lagom Tagger to Akka's new tagging function.

New Play/Lagom should use Play's tagger (yet to be defined).

Reply:
Actually, as long as users want to use TopicProducers or Read-Side Processors in Lagom, they have to stick with the current Lagom tagging strategy. 

Lagom requires the Event to be an `AggregateEvent` that has a `aggregateTag: AggregateEventTagger[E]` field. This is then later used internally to define the tagging function. 

We can only provide a new tagger utility when we have akka-projections. Then users can choose to use it and step-out of `AggregateEvent`